### PR TITLE
Small improvements

### PR DESCRIPTION
**Pull request by russcam**
- Avoid calling span.End() twice in DbSpanCommon
- Swap and assign new list for breakdown metrics

### CODE REVIEW

File: src/Elastic.Apm/Model/DbSpanCommon.cs [hunk 1]
```diff
@@ -56,7 +56,6 @@ internal void EndSpan(ISpan span, IDbCommand dbCommand, Outcome outcome, TimeSpa
 				}
 
 				capturedSpan.Outcome = outcome;
-				capturedSpan.End();
```
Comment (gregkalapos) [inline src/Elastic.Apm/Model/DbSpanCommon.cs:59]: Huhh, good catch - this was bad.

File: src/Elastic.Apm/Metrics/MetricsProvider/BreakdownMetricsProvider.cs [hunk 2]
```diff
@@ -84,30 +86,31 @@ public void CaptureTransaction(Transaction transaction)
...
 		{
-			var retVal = new List<MetricSet>(_itemsToSend.Count < 1000 ? _itemsToSend.Count : 1000);
+			List<MetricSet> retVal;
 
 			lock (_lock)
 			{
-				retVal.AddRange(_itemsToSend);
-				_itemsToSend.Clear();
+				retVal = _itemsToSend;
+				_itemsToSend = new List<MetricSet>(MetricLimit);
```
Comment (gregkalapos) [inline src/Elastic.Apm/Metrics/MetricsProvider/BreakdownMetricsProvider.cs:111]: I understand the change - but do we know if this is really better (in terms of perf. or something else)? Did you measure this? 

I'd not be able to tell which one is better - at least to me it's not obvious. Similar discussion [here](https://stackoverflow.com/questions/10901020/which-is-faster-clear-collection-or-instantiate-new).


Overall I'm fine with the change, I just have the impression that this doesn't really matter and I wonder what your thinking was.

**Review by gregkalapos — APPROVED** [review]
LGTM, just one small question below.

File: src/Elastic.Apm/Metrics/MetricsProvider/BreakdownMetricsProvider.cs (hunk 2 above)
Comment (russcam) [reply src/Elastic.Apm/Metrics/MetricsProvider/BreakdownMetricsProvider.cs:111]: My thinking is that a list allocation, copy items, and clear items would be slower than a list allocation and reassignment for larger count cases. The linked discussion is related, though it looks to be focused only on the original list, without concern for what happens to the existing items in the list, which is important in our use case, as we need to return as the samples.

I wrote a benchmark to measure this for the 1000 metrics case, and the assignment is frequently faster. I've extended it out to 10, 100, 500, 1000 count of metrics cases

```c#
using System;
using System.Collections.Generic;
using System.Linq;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Jobs;
using BenchmarkDotNet.Running;

namespace AddRangeBenchmark
{
    public class Program
    {
        public static void Main()
        {
            var summary = BenchmarkRunner.Run<BreakdownMetrics>();
        }

        [SimpleJob(RuntimeMoniker.Net461)]
        [SimpleJob(RuntimeMoniker.NetCoreApp30)]
        [SimpleJob(RuntimeMoniker.Net50)]
        public class BreakdownMetrics
        {
            private const int FixedCount = 1000;
            
            [Params(10, 100, 500, 1000)]
            public int Count { get; set; }
	
            private List<MetricSet> Samples()
            {
                var samples = new List<MetricSet>(FixedCount);
                for (int i = 0; i < Count; i++)
                {
                    samples.Add(new MetricSet(i,
                        new List<MetricSample>
                        {
                            new("span.self_time.count", i),
                            new("span.self_time.sum.us", i * 1000)
                        })
                    {
                        Span = new SpanInfo { Type = i.ToString(), SubType = i.ToString() },
                        Transaction = new TransactionInfo { Name = i.ToString(), Type = i.ToString() }
                    });
                }
                return samples;
            }


            [Benchmark]
            public (List<MetricSet>, List<MetricSet>) AddRangeCopy()
            {
                var samples = Samples();
                var retVal = new List<MetricSet>(samples.Count);
                retVal.InsertRange(0, samples);
                samples.Clear();
                return (retVal, samples);
            }

            [Benchmark]
            public (List<MetricSet>, List<MetricSet>) Assignment()
            {
                var samples = Samples();
                var retVal = samples;
                samples = new List<MetricSet>(FixedCount);
                return (retVal, samples);
            }
        }

        public class MetricSample
        {
            public MetricSample(string key, double value)
                => KeyValue = new KeyValuePair<string, double>(key, value);

            internal KeyValuePair<string, double> KeyValue { get; }
        }

        public class MetricSet
        {
            public MetricSet(long timestamp, IEnumerable<MetricSample> samples)
                => (Timestamp, Samples) = (timestamp, samples);

            /// <inheritdoc />
            public IEnumerable<MetricSample> Samples { get; set; }

            /// <inheritdoc />
            public long Timestamp { get; set; }

            public TransactionInfo Transaction { get; set; }

            public SpanInfo Span { get; set; }
        }

        public class TransactionInfo
        {
            public string Name { get; set; }
            public string Type { get; set; }
        }

        public class SpanInfo
        {
            public string Type { get; set; }
            public string SubType { get; set; }
        }
    }
}
```

The results are pretty mixed

```
// * Summary *

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19043.1165 (21H1/May2021Update)
Intel Core i7-8850H CPU 2.60GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET SDK=6.0.100-preview.4.21255.9
  [Host]               : .NET 5.0.7 (5.0.721.25508), X64 RyuJIT
  .NET 5.0             : .NET 5.0.7 (5.0.721.25508), X64 RyuJIT
  .NET Core 3.0        : .NET Core 3.0.3 (CoreCLR 4.700.20.6603, CoreFX 4.700.20.6701), X64 RyuJIT
  .NET Framework 4.6.1 : .NET Framework 4.8 (4.8.4400.0), X64 RyuJIT


|       Method |                  Job |              Runtime | Count |       Mean |      Error |     StdDev |     Median |
|------------- |--------------------- |--------------------- |------ |-----------:|-----------:|-----------:|-----------:|
| AddRangeCopy |             .NET 5.0 |             .NET 5.0 |    10 |   2.395 us |  0.0505 us |  0.1458 us |   2.367 us |
|   Assignment |             .NET 5.0 |             .NET 5.0 |    10 |   3.988 us |  0.0875 us |  0.2540 us |   3.885 us |
| AddRangeCopy |        .NET Core 3.0 |        .NET Core 3.0 |    10 |   2.866 us |  0.0622 us |  0.1814 us |   2.823 us |
|   Assignment |        .NET Core 3.0 |        .NET Core 3.0 |    10 |   5.473 us |  0.1090 us |  0.3145 us |   5.346 us |
| AddRangeCopy | .NET Framework 4.6.1 | .NET Framework 4.6.1 |    10 |   4.410 us |  0.0883 us |  0.2386 us |   4.328 us |
|   Assignment | .NET Framework 4.6.1 | .NET Framework 4.6.1 |    10 |   6.197 us |  0.1239 us |  0.3265 us |   6.131 us |
| AddRangeCopy |             .NET 5.0 |             .NET 5.0 |   100 |  18.830 us |  0.3754 us |  0.7410 us |  18.702 us |
|   Assignment |             .NET 5.0 |             .NET 5.0 |   100 |  20.827 us |  0.4067 us |  0.7015 us |  20.588 us |
| AddRangeCopy |        .NET Core 3.0 |        .NET Core 3.0 |   100 |  25.563 us |  0.5413 us |  1.5530 us |  25.284 us |
|   Assignment |        .NET Core 3.0 |        .NET Core 3.0 |   100 |  28.746 us |  0.5701 us |  0.7991 us |  28.665 us |
| AddRangeCopy | .NET Framework 4.6.1 | .NET Framework 4.6.1 |   100 |  39.558 us |  0.7791 us |  1.9402 us |  38.991 us |
|   Assignment | .NET Framework 4.6.1 | .NET Framework 4.6.1 |   100 |  39.934 us |  0.7418 us |  1.6125 us |  39.372 us |
| AddRangeCopy |             .NET 5.0 |             .NET 5.0 |   500 | 136.029 us |  2.2095 us |  1.7250 us | 136.166 us |
|   Assignment |             .NET 5.0 |             .NET 5.0 |   500 | 143.392 us |  2.8109 us |  2.3473 us | 143.287 us |
| AddRangeCopy |        .NET Core 3.0 |        .NET Core 3.0 |   500 | 230.297 us |  4.2190 us |  7.6078 us | 228.263 us |
|   Assignment |        .NET Core 3.0 |        .NET Core 3.0 |   500 | 234.353 us |  2.7036 us |  2.2576 us | 234.588 us |
| AddRangeCopy | .NET Framework 4.6.1 | .NET Framework 4.6.1 |   500 | 298.817 us |  5.5677 us |  4.6493 us | 298.656 us |
|   Assignment | .NET Framework 4.6.1 | .NET Framework 4.6.1 |   500 | 293.580 us |  3.4081 us |  2.6608 us | 293.649 us |
| AddRangeCopy |             .NET 5.0 |             .NET 5.0 |  1000 | 448.343 us |  7.5398 us | 12.8031 us | 442.990 us |
|   Assignment |             .NET 5.0 |             .NET 5.0 |  1000 | 458.874 us |  8.9047 us | 10.6005 us | 457.396 us |
| AddRangeCopy |        .NET Core 3.0 |        .NET Core 3.0 |  1000 | 530.470 us | 10.3587 us |  8.0874 us | 528.103 us |
|   Assignment |        .NET Core 3.0 |        .NET Core 3.0 |  1000 | 551.906 us |  9.8992 us |  8.2663 us | 553.515 us |
| AddRangeCopy | .NET Framework 4.6.1 | .NET Framework 4.6.1 |  1000 | 677.340 us | 11.7282 us |  9.1566 us | 674.395 us |
|   Assignment | .NET Framework 4.6.1 | .NET Framework 4.6.1 |  1000 | 654.919 us |  6.7451 us |  5.6325 us | 654.856 us |
```

I was seeing the 1000 metrics case as frequently faster across .NET 5.0, .NET Core 3.0 and .NET Framework 4.6.1, though the results above are faster for .NET Framework 4.6.1 only for the 1000 and 500 metrics case, in all other cases the AddRange and Clear is faster. Given these results, I'll leave as is 👍

**Pull request merged**